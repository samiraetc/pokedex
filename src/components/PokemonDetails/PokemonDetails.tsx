import { useState } from 'react';
import { Pokemon } from '../../types/typing';
import { removeFavoritePokemon, saveFavoritesPokemon, verifyIfIsFavorited } from '../../utils/favorites';
import Pokemons from '../Pokemon/Pokemon';
import ProgressBar from '../ProgressBar/ProgressBar';
import { MdOutlineCatchingPokemon, MdCatchingPokemon } from "react-icons/md";
import PokemonType from '../PokemonType/PokemonType';
import { FiArrowLeft } from "react-icons/fi";
import { Link } from 'react-router-dom';

export const COMPONENT_ID = 'pokedex-pokemon-details'

interface PokemonDetailsProps {
    pokemon: Pokemon
}

const PokemonDetails = ({ pokemon }: PokemonDetailsProps) => {
    const [areInFavorite, setAreInFavorite] = useState(verifyIfIsFavorited(pokemon.id))

    const handleClickToFavorite = () => {
        areInFavorite ? (
            removeFavoritePokemon(pokemon.id),
            setAreInFavorite(verifyIfIsFavorited(pokemon.id))
        ) : (
            saveFavoritesPokemon(pokemon),
            setAreInFavorite(verifyIfIsFavorited(pokemon.id)));
    }

    return (
        <div className="pokemon-details-container" id={COMPONENT_ID}>
            <div className='pokemon-details-card'>
                <Link to={`/`} >
                    <FiArrowLeft size={30} color="#373E59" />
                </Link>
                <div className='pokemon-details-favorite'>
                    <p className='pokemon-details-name'>{pokemon.name}</p>
                    <div onClick={() => handleClickToFavorite()} id={`${COMPONENT_ID}__favorite`}>
                        {areInFavorite ? (<MdCatchingPokemon size={30} color="#C22E28" />) : (<MdOutlineCatchingPokemon size={30} color="#373E59" />)}

                    </div>
                </div>

                <div className='pokemon-details-about'>

                    <div>
                        <Pokemons pokemon={pokemon} />
                        <div className="pokemon-details-type">
                            Types:
                            {pokemon.types.map((types: any, index: number) => {
                                const type = types.type.name;
                                return <PokemonType type={type} key={index} />
                            })}
                        </div>

                        <div className="pokemon-details-abilities">
                            Abilities:
                            {pokemon.abilities.map((value: any, index: number) => {
                                return <p key={index} className="pokemon-details-abilities__item">{value.ability.name}</p>
                            })}
                        </div>
                    </div>
                    <div className="pokemon-details-stats">
                        {pokemon.stats.map((value: any) => {
                            return (
                                <div key={pokemon.id} className='pokemon-details-stats__item'>
                                    {value.stat.name}
                                    <ProgressBar key={pokemon.id} completed={value.base_stat} />
                                </div>
                            )
                        })}
                    </div>
                </div>
            </div>

        </div>
    )
}

export default PokemonDetails;