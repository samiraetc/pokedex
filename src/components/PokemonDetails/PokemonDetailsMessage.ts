
export const PokemonDetailsMessages = {
    POKEMON_TYPES: 'Types:',
    POKEMON_ABILITIES: 'Abilities:'
}