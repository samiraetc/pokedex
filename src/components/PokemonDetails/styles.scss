.pokemon-details {
    &-card {
        position: relative;
        width: 800px;
        height: 100%;
        vertical-align: middle;

        .pokemon-avatar {
            width: 300px;
        }

    }


    &-container {
        display: flex;
        align-items: center;
        justify-content: center;
        margin: $pokedex-spacing-20;
    }

    &-type {
        margin: $pokedex-spacing-8;
        display: flex;
        align-items: center;
        justify-content: center;
        font-size: $pokedex-font-size-14;
        color: $pokedex-color-blue-strong;
        gap: $pokedex-spacing-20;

        &__item {
            font-weight: $pokedex-font-weight-normal;
            color: $pokedex-color-white;
            padding: $pokedex-spacing-8;
            border-radius: 5px;
        }
    }

    &-abilities {
        margin: $pokedex-spacing-8;
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: center;
        font-size: $pokedex-font-size-14;
        color: $pokedex-color-blue-strong;
        gap: $pokedex-spacing-8;

        &__item {
            font-weight: $pokedex-font-weight-normal;
        }
    }

    &-stats {
        margin: $pokedex-spacing-8;
        font-size: $pokedex-font-size-14;
        color: $pokedex-color-blue-strong;
        gap: $pokedex-spacing-8;
        display: flex;
        flex-direction: column;
        flex-wrap: wrap;
        width: 50%;
        padding: $pokedex-spacing-8;


        &__item {
            font-weight: $pokedex-font-weight-normal;
            text-align: center;
        }
    }

    &-about {
        display: flex;
        flex-direction: row;
        flex-wrap: wrap;
        align-items: center;
        justify-content: space-evenly;
        width: 100%;
    }

    &-name {
        color: $pokedex-color-blue-strong;
        font-size: $pokedex-font-size-40;
        text-align: center;
    }

}