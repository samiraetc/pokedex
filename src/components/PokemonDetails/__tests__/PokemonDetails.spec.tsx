import { render, configure, fireEvent } from '@testing-library/react'
import PokemonDetails, { COMPONENT_ID } from '../PokemonDetails';

configure({
    testIdAttribute: 'id',
});


describe('[POKEMON]', () => {

    test('should render Pokemon', () => {
        const { asFragment, getByText, getAllByText, getByTestId } = render(<PokemonDetails pokemon={{
            name: 'raticate',
            id: 20,
            abilities: [{ ability: { name: 'run-away' } }
            ],
            sprites: { front_default: 'string' },
            types: [{ type: { name: 'normal' } }],
            stats: [{ base_stat: 20, stat: { name: 'hp' } }],
            weight: 185,
            height: 7,
        }} />);

        expect(getAllByText('raticate'));
        expect(getByText('#20'));
        expect(getByTestId(COMPONENT_ID)).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test('should render Pokemon and click to favorite', () => {
        const { asFragment, getByText, getAllByText, getByTestId } = render(<PokemonDetails pokemon={{
            name: 'raticate',
            id: 20,
            abilities: [{ ability: { name: 'run-away' } }
            ],
            sprites: { front_default: 'string' },
            types: [{ type: { name: 'normal' } }],
            stats: [{ base_stat: 20, stat: { name: 'hp' } }],
            weight: 185,
            height: 7,
        }} />);

        expect(getAllByText('raticate'));
        fireEvent.click(getByTestId(`${COMPONENT_ID}__favorite`));
        expect(getByTestId(COMPONENT_ID)).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test('should render Pokemon and click to remove favorite', () => {
        const { asFragment, getByText, getAllByText, getByTestId } = render(<PokemonDetails pokemon={{
            name: 'raticate',
            id: 20,
            abilities: [{ ability: { name: 'run-away' } }
            ],
            sprites: { front_default: 'string' },
            types: [{ type: { name: 'normal' } }],
            stats: [{ base_stat: 20, stat: { name: 'hp' } }],
            weight: 185,
            height: 7,
        }} />);

        expect(getAllByText('raticate'));
        fireEvent.click(getByTestId(`${COMPONENT_ID}__favorite`));
        fireEvent.click(getByTestId(`${COMPONENT_ID}__favorite`));
        expect(getByTestId(COMPONENT_ID)).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

});