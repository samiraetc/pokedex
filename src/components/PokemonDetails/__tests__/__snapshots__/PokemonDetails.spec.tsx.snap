// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[POKEMON] should render Pokemon 1`] = `
<DocumentFragment>
  <div
    class="pokemon-details-container"
    id="pokedex-pokemon-details"
  >
    <div
      class="pokemon-details-card"
    >
      <div
        class="pokemon-details-favorite"
      >
        <p
          class="pokemon-details-name"
        >
          raticate
        </p>
        <div
          id="pokedex-pokemon-details__favorite"
        >
          <svg
            color="#373E59"
            fill="currentColor"
            height="30"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(55, 62, 89);"
            viewBox="0 0 512 512"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M458.4 64.3C400.6 15.7 311.3 23 256 79.3 200.7 23 111.4 15.6 53.6 64.3-21.6 127.6-10.6 230.8 43 285.5l175.4 178.7c10 10.2 23.4 15.9 37.6 15.9 14.3 0 27.6-5.6 37.6-15.8L469 285.6c53.5-54.7 64.7-157.9-10.6-221.3zm-23.6 187.5L259.4 430.5c-2.4 2.4-4.4 2.4-6.8 0L77.2 251.8c-36.5-37.2-43.9-107.6 7.3-150.7 38.9-32.7 98.9-27.8 136.5 10.5l35 35.7 35-35.7c37.8-38.5 97.8-43.2 136.5-10.6 51.1 43.1 43.5 113.9 7.3 150.8z"
            />
          </svg>
        </div>
      </div>
      <div
        class="pokemon-details-about"
      >
        <div>
          <div
            class="pokemon"
            id="pokedex-pokemon"
          >
            <img
              alt="raticate"
              class="pokemon-avatar"
              src="string"
            />
            <div
              class="pokemon__infos"
            >
              <p
                class="pokemon__name"
              >
                raticate
              </p>
              <p
                class="pokemon__number"
              >
                #20
              </p>
            </div>
            <div
              class="pokemon__infos"
            >
              <p>
                18.5 kg
              </p>
              <p>
                0.7 m
              </p>
            </div>
          </div>
          <div
            class="pokemon-details-type"
          >
            Types:
            <p
              class="pokemon-details-type__item"
              id="pokemon-details-type"
              style="background-color: rgb(168, 167, 122);"
            >
              normal
            </p>
          </div>
          <div
            class="pokemon-details-abilities"
          >
            Abilities:
            <p
              class="pokemon-details-abilities__item"
            >
              run-away
            </p>
          </div>
        </div>
        <div
          class="pokemon-details-stats"
        >
          <div
            class="pokemon-details-stats__item"
          >
            hp
            <div
              class="progress-bar"
              id="pokedex-progress-bar"
            >
              <div
                class="progress-bar__container"
              >
                <div
                  class="progress-bar__filler"
                  style="width: 20%;"
                />
              </div>
              <span
                class="progress-bar__label"
              >
                20
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`[POKEMON] should render Pokemon and click to favorite 1`] = `
<DocumentFragment>
  <div
    class="pokemon-details-container"
    id="pokedex-pokemon-details"
  >
    <div
      class="pokemon-details-card"
    >
      <div
        class="pokemon-details-favorite"
      >
        <p
          class="pokemon-details-name"
        >
          raticate
        </p>
        <div
          id="pokedex-pokemon-details__favorite"
        >
          <svg
            color="#C22E28"
            fill="currentColor"
            height="30"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(194, 46, 40);"
            viewBox="0 0 512 512"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
            />
          </svg>
        </div>
      </div>
      <div
        class="pokemon-details-about"
      >
        <div>
          <div
            class="pokemon"
            id="pokedex-pokemon"
          >
            <img
              alt="raticate"
              class="pokemon-avatar"
              src="string"
            />
            <div
              class="pokemon__infos"
            >
              <p
                class="pokemon__name"
              >
                raticate
              </p>
              <p
                class="pokemon__number"
              >
                #20
              </p>
            </div>
            <div
              class="pokemon__infos"
            >
              <p>
                18.5 kg
              </p>
              <p>
                0.7 m
              </p>
            </div>
          </div>
          <div
            class="pokemon-details-type"
          >
            Types:
            <p
              class="pokemon-details-type__item"
              id="pokemon-details-type"
              style="background-color: rgb(168, 167, 122);"
            >
              normal
            </p>
          </div>
          <div
            class="pokemon-details-abilities"
          >
            Abilities:
            <p
              class="pokemon-details-abilities__item"
            >
              run-away
            </p>
          </div>
        </div>
        <div
          class="pokemon-details-stats"
        >
          <div
            class="pokemon-details-stats__item"
          >
            hp
            <div
              class="progress-bar"
              id="pokedex-progress-bar"
            >
              <div
                class="progress-bar__container"
              >
                <div
                  class="progress-bar__filler"
                  style="width: 20%;"
                />
              </div>
              <span
                class="progress-bar__label"
              >
                20
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`[POKEMON] should render Pokemon and click to remove favorite 1`] = `
<DocumentFragment>
  <div
    class="pokemon-details-container"
    id="pokedex-pokemon-details"
  >
    <div
      class="pokemon-details-card"
    >
      <div
        class="pokemon-details-favorite"
      >
        <p
          class="pokemon-details-name"
        >
          raticate
        </p>
        <div
          id="pokedex-pokemon-details__favorite"
        >
          <svg
            color="#C22E28"
            fill="currentColor"
            height="30"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(194, 46, 40);"
            viewBox="0 0 512 512"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
            />
          </svg>
        </div>
      </div>
      <div
        class="pokemon-details-about"
      >
        <div>
          <div
            class="pokemon"
            id="pokedex-pokemon"
          >
            <img
              alt="raticate"
              class="pokemon-avatar"
              src="string"
            />
            <div
              class="pokemon__infos"
            >
              <p
                class="pokemon__name"
              >
                raticate
              </p>
              <p
                class="pokemon__number"
              >
                #20
              </p>
            </div>
            <div
              class="pokemon__infos"
            >
              <p>
                18.5 kg
              </p>
              <p>
                0.7 m
              </p>
            </div>
          </div>
          <div
            class="pokemon-details-type"
          >
            Types:
            <p
              class="pokemon-details-type__item"
              id="pokemon-details-type"
              style="background-color: rgb(168, 167, 122);"
            >
              normal
            </p>
          </div>
          <div
            class="pokemon-details-abilities"
          >
            Abilities:
            <p
              class="pokemon-details-abilities__item"
            >
              run-away
            </p>
          </div>
        </div>
        <div
          class="pokemon-details-stats"
        >
          <div
            class="pokemon-details-stats__item"
          >
            hp
            <div
              class="progress-bar"
              id="pokedex-progress-bar"
            >
              <div
                class="progress-bar__container"
              >
                <div
                  class="progress-bar__filler"
                  style="width: 20%;"
                />
              </div>
              <span
                class="progress-bar__label"
              >
                20
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
